src/py_zfs_vdev.c

```diff
@@ -0,0 +1,320 @@
+	if (strcmp(cstate, "UNAVAIL") == 0 || strcmp(cstate, "OFFLINE") == 0)
+		return;
+	Py_BEGIN_ALLOW_THREADS
+	ctype = fnvlist_lookup_string(tree, ZPOOL_CONFIG_TYPE);
+	Py_END_ALLOW_THREADS
+	if (strcmp(ctype, VDEV_TYPE_FILE) == 0)
+		return;
+	else if (strcmp(ctype, VDEV_TYPE_DISK) == 0) {
+		Py_BEGIN_ALLOW_THREADS
+		cpath = fnvlist_lookup_string(tree, ZPOOL_CONFIG_PATH);
+		Py_END_ALLOW_THREADS
+		path = PyUnicode_FromString(cpath);
+		PYZFS_ASSERT((PyList_Append(list, path) == 0),
+		    "Failed to add disk path to Python list.");
+		return;
+	} else {
+		Py_BEGIN_ALLOW_THREADS
+		found = (nvlist_lookup_nvlist_array(tree, ZPOOL_CONFIG_CHILDREN,
+		    &child, &children) == 0);
+		Py_END_ALLOW_THREADS
+		if (found) {
+			for (uint_t i = 0; i < children; ++i) {
+				vdev_get_disks_impl(list, child[i]);
+			}
+		}
+	}
+	return;
+}
+
+static
+PyObject *py_zfs_vdev_get_disks(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	PyObject *list;
+
+	list = PyList_New(0);
+	PYZFS_ASSERT(list, "Failed to create Python list object.");
+	vdev_get_disks_impl(list, v->vdev_tree);
+
+	return (list);
+}
+
+static
+PyGetSetDef zfs_vdev_getsetters[] = {
+	{
+		.name	= "type",
+		.get	= (getter)py_zfs_vdev_get_type
+	},
+	{
+		.name	= "path",
+		.get	= (getter)py_zfs_vdev_get_path
+	},
+	{ .name = NULL }
+};
+
+static
+PyMethodDef zfs_vdev_methods[] = {
+	{
+		.ml_name = "asdict",
+		.ml_meth = py_zfs_vdev_asdict,
+		.ml_flags = METH_VARARGS | METH_KEYWORDS
+	},
+	{
+		.ml_name = "name",
+		.ml_meth = py_zfs_vdev_get_name,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "guid",
+		.ml_meth = py_zfs_vdev_get_guid,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "status",
+		.ml_meth = py_zfs_vdev_get_status,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "size",
+		.ml_meth = py_zfs_vdev_get_size,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "children",
+		.ml_meth = py_zfs_vdev_get_children,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "disks",
+		.ml_meth = py_zfs_vdev_get_disks,
+		.ml_flags = METH_NOARGS,
+	},
+	{ NULL, NULL, 0, NULL }
+};
+
+PyTypeObject ZFSVdev = {
+	.tp_name = "ZFSVdev",
+	.tp_basicsize = sizeof (py_zfs_vdev_t),
+	.tp_methods = zfs_vdev_methods,
+	.tp_getset = zfs_vdev_getsetters,
+	.tp_new = py_zfs_vdev_new,
+	.tp_init = py_zfs_vdev_init,
+	.tp_doc = "ZFSVdev",
+	.tp_dealloc = (destructor)py_zfs_vdev_dealloc,
+	.tp_repr = py_repr_zfs_vdev,
+	.tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE
+};
+
+py_zfs_vdev_t *init_zfs_vdev(py_zfs_t *lzp, zpool_handle_t *zhp,
+    nvlist_t *tree, PyObject *parent)
+{
+	py_zfs_vdev_t *out = NULL;
+	const char *type;
+	const char *path;
+	boolean_t path_found;
+
+	out = (py_zfs_vdev_t *)PyObject_CallFunction((PyObject *)&ZFSVdev, NULL);
+	PYZFS_ASSERT(out, "Failed to create ZFSVdev type object.");
+
+	out->pylibzfsp = lzp;
+	Py_INCREF(lzp);
+	out->parent = parent;
+	Py_XINCREF(parent);
+	out->zhp = zhp;
+	out->vdev_tree = tree;
+
+	Py_BEGIN_ALLOW_THREADS
```

Comment:
Maybe add comment here that fnvlist can't fail (for posterity sake).

Reply:
I have added the comment.